mitxgraders/comparers/comparers.py

```diff
@@ -112,30 +113,108 @@ class EqualityComparer(Comparer):
-            shape = tuple() if isinstance(expected_input, Number) else expected_input.shape
+            shape = tuple() if isinstance(expected_eval, Number) else expected_eval.shape
             utils.validate_shape(student_eval, shape)
 
-        # If provided, apply the transform function
-        if self.config['transform'] is not None:
-            expected_input = self.config['transform'](expected_input)
-            student_eval = self.config['transform'](student_eval)
+    def transform(self, evaluation):
+        f = self.config['transform']
+        return evaluation if f is None else f(evaluation)
+
 
-        return utils.within_tolerance(expected_input, student_eval)
+    def __call__(self, comparer_params_eval, student_eval, utils):
+        expected_eval = comparer_params_eval[0]
+        self.validate(expected_eval, student_eval, utils)
+
+        expected_eval = self.transform(expected_eval)
+        student_eval = self.transform(student_eval)
+
+        return utils.within_tolerance(expected_eval, student_eval)
 
 equality_comparer = EqualityComparer()
 
+class MatrixEntryComparer(EqualityComparer):
+    """
+    Default comparer for MatrixGrader. Compares student and instructor matirx
+    evaluations entry-by-entry for equality.
+
+    Configuration
+    =============
+
+        transform (None | function): same as EqualityComparer, defaults to None
+        entry_partial_credit ('proportional'|number): Determines how partial credit
+            is awarded. If set to 'proportional', then credit is proportional to
+            the number of correct matrix entries. If a numeric value is provided,
+            this flat rate of partial credit is provided as long as some but
+            not all entries are correct.
+
+            Default is the numeric value 0 (no partial credit).
+        entry_partial_msg (str): A text string message shown when partial credit
+            is awarded. The string may optionally contain the formatting key {error_indices},
+            which will be replaced with the indices of the incorrect matrix entries.
+    """
+
+    default_msg = "Matrix entries at indices {error_indices} are incorrect."
+    schema_config = EqualityComparer.schema_config.extend({
+        Required('entry_partial_credit', default=0): Any(All(Number, Range(0, 1)), 'proportional'),
+        Required('entry_partial_msg', default=default_msg): text_string
+    })
+
+    @staticmethod
+    def format_message_with_locations(format_string, locs):
+        """
+        Returns format_string with {error_indices} replaced by 1-indexed error
+        locations.
+
+        Arguments:
+            format_string: a string that may contain {error_indices} formatting key.
+            locs: a 0-indexed numpy array of error locations
+        """
+        locs_as_str = ", ".join(map(six.text_type, locs+1))
+        return format_string.format(error_indices=locs_as_str)
+
+    def __call__(self, comparer_params_eval, student_eval, utils):
```

Comment:
This needs to be done with a correlated comparer rather than a simple comparer that computes the truth table across all samples. Otherwise, it's quite possible that the percentage and message are incorrect (especially when using integer sampling and Kronecker delta functions).